nautobot/core/tasks.py

```diff
@@ -53,3 +54,20 @@ def get_releases(pre_releases=False):
 
     # Since this is a Celery task, we can't return Version objects as they are not JSON serializable.
     return [(str(version), url) for version, url in releases]
+
+
+@celery.nautobot_task(ignore_result=True)
+def handle_time_limit_exceeded(task_id):
+    """
+    Handler for when a running `Job` hits its hard `time_limit` to ensure that
+    the `JobResult` is marked as a failure and the traceback is saved.
+    """
+    from nautobot.extras.choices import JobResultStatusChoices
+    from nautobot.extras.models import JobResult
+
+    jr = JobResult.objects.get(task_id=task_id)
+    # TODO(jathan): Assert this is the correct accessor for `time_limit`
+    time_limit = jr.job_model.time_limit
+    jr.set_status(JobResultStatusChoices.STATUS_FAILURE)
+    jr.traceback = str(TimeLimitExceeded(time_limit))
```

Comment:
Can we get the actual traceback in some fashion? Would be really helpful to know where in the task/job the time limit was hit, for example.

Reply:
Here is what the logs look like:

```
[2023-02-06 16:06:20,128: WARNING/MainProcess] Soft time limit (0.5s) exceeded for nautobot.extras.jobs.run_job[3796a944-7220-45d4-9ad5-604f2f62ecde]
16:06:20.178 WARNING nautobot.jobs.BunkJob jobs.py                                    log() :
  Time's up! About to crash.
[2023-02-06 16:06:20,178: WARNING/ForkPoolWorker-8] Time's up! About to crash.
  Performing nautobot serialization on <nautobot.core.utils.requests.NautobotFakeRequest object at 0x126337eb0> for type nautobot.core.utils.requests.NautobotFakeRequest
[2023-02-06 16:06:20,670: ERROR/MainProcess] Task handler raised error: TimeLimitExceeded(1.0)
Traceback (most recent call last):
  File "/Users/jathan/Library/Caches/pypoetry/virtualenvs/nautobot-b2ttWva6-py3.9/lib/python3.9/site-packages/billiard/pool.py", line 684, in on_hard_timeout
    raise TimeLimitExceeded(job._timeout)
billiard.exceptions.TimeLimitExceeded: TimeLimitExceeded(1.0,)
[2023-02-06 16:06:20,677: ERROR/MainProcess] Hard time limit (1.0s) exceeded for nautobot.extras.jobs.run_job[3796a944-7220-45d4-9ad5-604f2f62ecde]
  Performing nautobot serialization on <nautobot.core.utils.requests.NautobotFakeRequest object at 0x126337eb0> for type nautobot.core.utils.requests.NautobotFakeRequest
[2023-02-06 16:06:20,783: INFO/MainProcess] Task nautobot.core.tasks.handle_time_limit_exceeded[68b530db-a0a9-43db-8d75-a7302aabde5b] received
[2023-02-06 16:06:20,786: ERROR/MainProcess] Process 'ForkPoolWorker-8' pid:79723 exited with 'signal 9 (SIGKILL)'
[2023-02-06 16:06:20,841: INFO/ForkPoolWorker-1] Task nautobot.core.tasks.handle_time_limit_exceeded[68b530db-a0a9-43db-8d75-a7302aabde5b] succeeded in 0.05540520799999982s: None
```

- First there is the `SoftTimeLimitExceeded` which we catch and handle.
- Then there is a `TimeLimitExceeded` exception that is raised but it's caught by the `billiard.pool` code and that's what results in the process on the worker being called
- Therefore the context of the exception is lost as far as I can tell
- The `Request.on_timeout()` handler on the database backend makes a call to the [backend for the task to `mark_as_failure()`](https://github.com/celery/celery/blob/main/celery/worker/request.py#L528-L531) but best I can tell here, this call is never made because the process is terminated before it even gets here. Could merely be a timing issue to the database connection, but it requires more R&D to get to the bottom of it.

So... I figured having the exception that resulted in the termination is better than nothing.